lib/stream.js

```diff
@@ -115,49 +115,49 @@ const waitForCustomStreamsEnd = (stdioStreamsGroups, waitForStream) => stdioStre
-const pEvent = (eventEmitter, eventName, {signal}) => new Promise(resolve => {
-	const listener = (...payload) => {
-		resolve([eventName, ...payload]);
-	};
-
-	eventEmitter.once(eventName, listener);
-	addAbortListener(signal, () => {
-		eventEmitter.removeListener(eventName, listener);
-	});
-});
-
 // Wraps `finished(stream)` to handle the following case:
 //  - When the child process exits, Node.js automatically calls `childProcess.stdin.destroy()`, which we need to ignore.
 //  - However, we still need to throw if `childProcess.stdin.destroy()` is called before child process exit.
-const onFinishedStream = async ([originalStdin], processExitPromise, stream) => {
-	const finishedPromise = finished(stream, {cleanup: true});
-	if (stream !== originalStdin) {
-		await finishedPromise;
-		return;
-	}
-
-	try {
-		await finishedPromise;
-	} catch {
-		await Promise.race([processExitPromise, finishedPromise]);
-	}
+const onFinishedStream = async ([originalStdin], exitPromise, stream) => {
+	await Promise.race([
+		...(stream === originalStdin ? [exitPromise] : []),
+		finished(stream, {cleanup: true}),
+	]);
 };
 
-const throwOnProcessError = async processErrorPromise => {
-	const [, error] = await processErrorPromise;
+const throwOnProcessError = async (spawned, {signal}) => {
+	const [error] = await once(spawned, 'error', {signal});
 	throw error;
 };
 
-// First the `spawn` event is emitted, then `exit`.
-// If the `error` event is emitted:
-//  - before `spawn`: `exit` is never emitted.
-//  - after `spawn`: `exit` is always emitted.
-// We only want to listen to `exit` if it will be emitted, i.e. if `spawn` has been emitted.
-// Therefore, the arguments order of `Promise.race()` is significant.
-const waitForFailedProcess = async (processSpawnPromise, processErrorPromise, processExitPromise) => {
-	const [eventName] = await Promise.race([processSpawnPromise, processErrorPromise]);
-	return eventName === 'spawn' ? processExitPromise : [];
+// If `error` is emitted before `spawn`, `exit` will never be emitted.
+// However, `error` might be emitted after `spawn`, e.g. with the `signal` option.
+// In that case, `exit` will still be emitted.
+// Since the `exit` event contains the signal name, we want to make sure we are listening for it.
+// This function also takes into account the following unlikely cases:
+//  - `exit` being emitted in the same microtask as `spawn`
+//  - `error` being emitted multiple times
+const waitForExit = async spawned => {
+	const [spawnPayload, exitPayload] = await Promise.allSettled([
+		once(spawned, 'spawn'),
+		once(spawned, 'exit'),
+	]);
+
+	if (spawnPayload.status === 'rejected') {
+		return [];
+	}
+
+	return exitPayload.status === 'rejected'
+		? waitForProcessExit(spawned)
+		: exitPayload.value;
+};
+
+const waitForProcessExit = async spawned => {
+	try {
+		return await once(spawned, 'exit');
+	} catch {
+		return waitForProcessExit(spawned);
```

Comment:
Once `spawn` has been emitted, `exit` is guaranteed to be emitted, even if `error` is emitted. 
I.e. no possible infinite recursion.